### PR TITLE
TW-346: Add highlight in search

### LINKED ISSUE #346: [Search] Hightlight result in search for contacts & chats
### Desc:

- Need to highlight the keyword inside some fields include this keyword 

### DoD:

- Highlight title (`displayName`) and subtitle (`matrixId`)

Image

### PR DESCRIPTION
**Pull request by imGok**
- Issue: #346 

1. List recent contacts is empty

https://github.com/linagora/twake-on-matrix/assets/99852347/c30e7ac0-a8ca-42ba-9812-90a3122c2207

2. Searching and scrolling

https://github.com/linagora/twake-on-matrix/assets/99852347/240d8c6d-64d5-4a9d-916d-8854c97e62cf

### CODE REVIEW

**Comment by hoangdat** [comment]
Thank @imGok, but in this screen we will not have highlight for the text. This ticket aims for the `feature/search`

File: lib/pages/search/recent_item_widget.dart [hunk 1]
```diff
@@ -58,18 +59,23 @@ class RecentItemWidget extends StatelessWidget {
...
                             overflow: TextOverflow.ellipsis,
                             letterSpacing: 0.15,
                             color: Theme.of(context).colorScheme.onSurfaceVariant,
                           ),
                         ),
+                        highlightStyle: Theme.of(context).textTheme.titleMedium?.merge(
+                          TextStyle(
+                            overflow: TextOverflow.ellipsis,
+                            letterSpacing: 0.15,
+                            color: Colors.amber[900],
```
Comment (nqhhdev) [inline lib/pages/search/recent_item_widget.dart:75]: IMO: Why don't you use M3 color ?

Comment (imGok) [reply lib/pages/search/recent_item_widget.dart:75]: WDYM ? Colors is a material class

**Review by drminh2807 — APPROVED** [review]

**Review by nqhhdev — APPROVED** [review]

File: lib/widgets/highlight_text.dart [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      RegExp(searchWord!, caseSensitive: false),
+      onMatch: (Match match) {
+        spans.add(buildHighlightSpan(match.group(0)!));
+        return '';
+      },
+      onNonMatch: (String nonMatch) {
+        spans.add(buildNormalSpan(nonMatch));
+        return '';
+      },
+    );
```
Comment (sherlockvn) [inline lib/widgets/highlight_text.dart:34]: I think you should put if in an extension, or at least in a method

Comment (imGok) [reply lib/widgets/highlight_text.dart:46]: done in 007a3586e85c17686fbfee3c87e4f2a010a9435e

File: lib/pages/search/recent_item_widget.dart [hunk 3]
```diff
@@ -58,18 +59,16 @@ class RecentItemWidget extends StatelessWidget {
                   child: Column(
                     crossAxisAlignment: CrossAxisAlignment.start,
                     children: [
-                      Text(
-                        presentationSearch.displayName ?? "",
-                        overflow: TextOverflow.ellipsis,
-                        maxLines: 1,
-                        softWrap: false,
+                      HighlightText(
```
Comment (dab246) [inline lib/pages/search/recent_item_widget.dart:62]: Good !

**Review by dab246 — APPROVED** [review]
lgtm

**Comment by nqhhdev** [comment]
Pls rebase your PR

**Comment by hoangdat** [comment]
why we have the blank space in the screen? Why not remove it when we don't have anything to show on it?

image

**Comment by nqhhdev** [comment]
> why we have the blank space in the screen? Why not remove it when we don't have anything to show on it?
> 
> image

1. I updated for this case. Pls check demo 
2. I found some bugs and then fixed

**Review by sherlockvn — APPROVED** [review]

File: lib/pages/search/recent_item_widget.dart [hunk 4]
```diff
@@ -167,19 +170,17 @@ class _ContactInformation extends StatelessWidget {
...
-          overflow: TextOverflow.ellipsis,
-          maxLines: 1,
-          softWrap: false,
-          style: Theme.of(context).textTheme.bodyMedium?.merge(
-            TextStyle(
-              overflow: TextOverflow.ellipsis,
-              letterSpacing: 0.15,
-              color: LinagoraRefColors.material().tertiary[30],
+          HighlightText(
+            text: presentationSearch.directChatMatrixID ?? "",
```
Comment (hoangdat) [inline lib/pages/search/recent_item_widget.dart:174]: how about display name? how about the email?

Comment (nqhhdev) [reply lib/pages/search/recent_item_widget.dart:174]: - It's information widget and Julian support highlights for this widget. If it was not fine I'll remove it. WDYT?

Simulator Screenshot - iPhone 14 - 2023-07-31 at 12 31 10

**Pull request merged**